### DISCUSSION TITLE
How to view Roslyn implementation of a class

### DISCUSSION THREAD

**Discussion by RobertGBryan** [dotnet/roslyn#55832] [Q&A] [answered]
I am quite new to Roslyn, so please bear with me:

1. Can I view the implementation for the `ArrayPool<T>` class under the dotnet / roslyn branch here on GitHub? If so, how?

I have tried using file finder. For example, when I enter ArrayPool, Array, Pool, or Buffers, it does not show any results for `ArrayPool<T>`. I have also tried doing a Google search, but that does not produce any results either.

**Comment by PathogenDavid** [comment, marked as answer]
Roslyn is the C# compiler, it sounds like you're looking for the implementation of the types provided by the .NET Runtime. That all lives over in https://github.com/dotnet/runtime

The easiest way to browse the source code is using the [.NET Source Browser](https://source.dot.net/), here's the implementation of [`System.Buffers.ArrayPool<T>`](https://source.dot.net/#System.Private.CoreLib/ArrayPool.cs,87992df74cbf00ad).

You'll find the actual implementation is provided by either `TlsOverPerCoreLockedStacksArrayPool<T>` or `ConfigurableArrayPool<T>` depending on where the pool is coming from. You can click on them in the source view to jump to their implementations, or click on the `ArrayPool<T>` class definition to get a list of implementations listed on the left.

You can also click the "Web Access" button at the bottom-right to jump to [the source on GitHub](https://github.com/dotnet/runtime/blob/3d5d26181cd7bc07ea6c6f87710c14ccd043b415/src/libraries/System.Private.CoreLib/src/System/Buffers/ArrayPool.cs).

(There's also a similar tool for the actual Roslyn code too if you ever find yourself wanting that: https://sourceroslyn.io/)

**Reply by RobertGBryan** [reply]
Nice. Thank you.
